File: solutions/number-of-provinces/number-of-provinces.cpp
// @Title: Number of Provinces
// @Author: colinjxc
// @Date: 2022-01-25T18:18:44+08:00
// @URL: https://leetcode-cn.com/problems/number-of-provinces


class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int N = isConnected.size();
        std::vector<bool> visited(N, false);

        std::function<void(int)> dfsVisit = [&](int n) {
            if (visited[n]) return;
            visited[n] = true;
            for (int i = 0; i < isConnected[n].size(); i++) {
                if (isConnected[n][i] == 1) {
                    dfsVisit(i);
                }     
            }
        };

        int cnt = 0;
        for (int i = 0; i < N; i++) {
            if (!visited[i]) {
                dfsVisit(i);
                cnt++;
            }
        }
        return cnt;
    }
};

File: solutions/number-of-provinces/number-of-provinces.go
// @Title: Number of Provinces
// @Author: colinjxc
// @Date: 2021-10-28T22:03:16+08:00
// @URL: https://leetcode-cn.com/problems/number-of-provinces


func findCircleNum(isConnected [][]int) int {
    pCnt := 0
    vis := make([]bool, len(isConnected))
    
    var dfs func(int)
    dfs = func(from int) {
        vis[from] = true;
        for i, c := range isConnected[from] {
            if c == 1 && !vis[i] {
                dfs(i)
            }
        }
    }

    for i, _:= range isConnected {
        if !vis[i] {
            pCnt++
            dfs(i)
        }
    }
    return pCnt
}
